File: socks.java
import java.awt.*;
import java.applet.*;
import java.awt.image.FilteredImageSource;
import java.awt.image.CropImageFilter;

public class socks extends klevel {
    int numMatched;
    String[] names = {"ein Diamant", "eine Socke", "eine Jacke", "eine Hose", "ein Schuh", "ein Säckchen", "ein Sporthemd", "ein Hut"};
    Image[] casePic = new Image[2];
    Image[] itemPic = new Image[8];
    int[][] boxID = new int[5][3];
    int[][] boxST = new int[5][3];
    Image cases, items, sockback;
    Rectangle[][] content = new Rectangle[5][3];
    boolean timOn = false;
    Point[] curOpen = new Point[2];
    int numOpen = 0, numPRINT = 0;
    int[][] PRINTed = new int[5][3];
    AudioClip Ja, Nein, DiamantSound;

    public socks() {
        inst = "Click and Match the Items.";
        plot = "Was ist im Koffer?";
    }

    public void loadData() {
        Ja = waitAudioClip("yes.au");
        Nein = waitAudioClip("no.au");
        DiamantSound = waitAudioClip("diamant.au");
        cases = waitImage("cases.gif");
        items = waitImage("items.gif");
        sockback = waitImage("sockback.jpg");
        MediaTracker track = new MediaTracker(this);
        for ( int buf= 0; buf < 2; buf++ )
        {
            casePic[buf] = cropImage(cases, buf*40, 0, 40, 40);
            track.addImage(casePic[buf], 0);
        }
        for ( int buf = 0; buf < 8; buf++) {
            itemPic[buf] = cropImage(items, buf*20, 0, 20, 20);
            track.addImage(itemPic[buf], 0);
        }


        try
        {
            track.waitForID(0);
        } catch (Exception e) { }



        int rx=0, ry=0;
        int ct = 0;

        for (int item=0; item< 8; item++)
        {

        	for (int howmany=0; howmany<((item==0)?1:2); howmany++)
        	{
        		do
        		{

        			rx = (int)(Math.random() * 5d);
        			ry = (int)(Math.random() * 3d);
        		} while (boxID[rx][ry]!=0);

        		boxID[rx][ry] = item;
        		ct++;

        	}
        }
        for( int i = 0; i < 5; i++) for (int j = 0; j < 3; j++) {
            content[i][j] = new Rectangle(i*40+20*(i)+60, j*40+20*(j)+100, 40, 40);

        }


            speed = 1000;
            start();
   }

    public void drawStatic (Graphics g)
    {
        g.drawImage(sockback,0,0,this);
    }


    public void drawNormal(Graphics g) {
        numMatched = 0;
            for (int i = 0; i < 5; i++) for ( int j = 0; j < 3; j++) {
                switch (boxST[i][j]) {
                    case 0:
                        g.drawImage(casePic[0], content[i][j].x, content[i][j].y, this);
                        break;
                    case 1:
                        g.drawImage(casePic[1], content[i][j].x, content[i][j].y, this);
                        g.drawImage(itemPic[boxID[i][j]], content[i][j].x+10, content[i][j].y, this);
                        if ((numPRINT == 0) && (PRINTed[i][j] == 0)) {PRINTed[i][j] = 1;} else
                        if ((numPRINT == 1) && (PRINTed[i][j] == 0)) { PRINTed[i][j] = 2;}

                        if (PRINTed[i][j] == 1) {
                            drawText(g, names[boxID[i][j]], new Point(25,25), RIGHT, TOP, Color.white, false);
                            numPRINT = 1;

                        } else if (PRINTed[i][j] == 2) {
                            drawText(g, "und",  new Point(200,25), CENTER, TOP, Color.white, false);
                            drawText(g, names[boxID[i][j]], new Point(375,25), LEFT, TOP, Color.white, false);
                            numPRINT = 0;

                        }
                        break;

                    case 2:
                        numMatched++;
                        break;
                    case 3:

                        drawText(g, "ein Diamant", new Point(200,75), CENTER, TOP, Color.cyan, false);
                        g.drawImage(casePic[1], content[i][j].x, content[i][j].y, this);
                        g.drawImage(itemPic[0], content[i][j].x+10, content[i][j].y, this);
                        boxST[i][j] = 2;
                        DiamantSound.play();
                        enableNextLevel();
                        changeScore(100);
                        timOn = false;

                        break;
                }



            }


                if (numOpen >= 2) {
                    numOpen = 0;
                    timOn = false;

                checkMatch(g, curOpen[0].x, curOpen[0].y, curOpen[1].x, curOpen[1].y);
                }

                if (timOn && numMatched < 15) { timOn = false; stop(); } else if (numMatched >= 15) {
                    displayText(g, "Next Level: +100pts");
                    changeScore(100);
                    stop();
                    numOpen =0;
} else  {timOn = true; }


    }
    public boolean mouseDown(Event e, int x, int y) {
        for ( int i = 0; i < 5; i++) for ( int j = 0; j < 3; j++) {



            if ((content[i][j].inside(x, y) == true) && (boxST[i][j] == 0)) {
                boxST[i][j] = 1;
                if (boxID[i][j] == 0) {

                boxST[i][j] = 3;



                        } else {

                numOpen++;
                curOpen[numOpen-1] = new Point(i , j);
                        }
                timOn = false;
                start();
            }

        }
    return true;
    }


    public boolean checkMatch(Graphics g, int x1, int y1, int x2, int y2) {

        boolean retval = false;
        if (boxID[x1][y1] == boxID[x2][y2]) {
            drawText(g, "Ja", new Point(200,75), CENTER, TOP, Color.yellow, false);
            Ja.play();
            boxST[x1][y1] = 2;
            boxST[x2][y2] = 2;
            timOn = false;
            retval = true;
            changeScore(50);
        } else {
            boxST[x1][y1] = 0;
            boxST[x2][y2] = 0;
            drawText(g, "Nein", new Point(200,75), CENTER, TOP, Color.yellow, false);
            Nein.play();
            timOn = false;
            changeScore(-10);
        }

            PRINTed[x1][y1] = 0;
            PRINTed[x2][y2] = 0;
numPRINT = 0;
        return retval;
    }
    public boolean drawCrap(Graphics g, int i, int j, int numPRINT) {
        switch (numPRINT) {
            case 0:
             drawText(g, names[boxID[i][j]], new Point(25,25), RIGHT, TOP, Color.white, false);
             numPRINT++;
             break;
             case 1:
             drawText(g, "und" , new Point(200, 25), CENTER, TOP, Color.white, false);
             drawText(g, names[boxID[i][j]], new Point(375, 25), LEFT, TOP, Color.white, false);
             numPRINT = 0;
             break;
        }

    return true;
    }


}

File: js/socks.js
"use strict";

// Port of socks.java (memory-match game). The original logic is quirky
// (including the diamond placement: item 0 marks its cell as 0, i.e. still
// "free", so the diamond ends up in whichever cell is left unfilled) and is
// preserved as-is.

class SocksLevel extends KLevel {
  constructor(game) {
    super(game);
    this.plot = "Was ist im Koffer?";
    this.inst = "Click and Match the Items.";

    this.names = ["ein Diamant", "eine Socke", "eine Jacke", "eine Hose",
                  "ein Schuh", "ein Säckchen", "ein Sporthemd", "ein Hut"];
    this.numMatched = 0;
    this.casePic = [];
    this.itemPic = [];
    this.boxID = Array.from({ length: 5 }, () => [0, 0, 0]);
    this.boxST = Array.from({ length: 5 }, () => [0, 0, 0]);
    this.content = Array.from({ length: 5 }, () => [null, null, null]);
    this.timOn = false;
    this.curOpen = [null, null];
    this.numOpen = 0;
    this.numPRINT = 0;
    this.PRINTed = Array.from({ length: 5 }, () => [0, 0, 0]);
  }

  async loadData() {
    this.Ja = new Snd("assets/yes.wav");
    this.Nein = new Snd("assets/no.wav");
    this.DiamantSound = new Snd("assets/diamant.wav");
    const cases = await loadImage("assets/cases.gif");
    const items = await loadImage("assets/items.gif");
    this.sockback = await loadImage("assets/sockback.jpg");

    for (let buf = 0; buf < 2; buf++) {
      this.casePic[buf] = cropImage(cases, buf * 40, 0, 40, 40);
    }
    for (let buf = 0; buf < 8; buf++) {
      this.itemPic[buf] = cropImage(items, buf * 20, 0, 20, 20);
    }

    for (let item = 0; item < 8; item++) {
      for (let howmany = 0; howmany < ((item === 0) ? 1 : 2); howmany++) {
        let rx, ry;
        do {
          rx = Math.floor(Math.random() * 5);
          ry = Math.floor(Math.random() * 3);
        } while (this.boxID[rx][ry] !== 0);
        this.boxID[rx][ry] = item;
      }
    }
    for (let i = 0; i < 5; i++) {
      for (let j = 0; j < 3; j++) {
        this.content[i][j] = { x: i * 60 + 60, y: j * 60 + 100, w: 40, h: 40 };
      }
    }

    this.speed = 1000;
    this.start();
  }

  drawStatic(g) {
    g.drawImage(this.sockback, 0, 0);
  }

  drawNormal(g) {
    g.font = "12px sans-serif";
    this.numMatched = 0;
    for (let i = 0; i < 5; i++) {
      for (let j = 0; j < 3; j++) {
        const c = this.content[i][j];
        switch (this.boxST[i][j]) {
          case 0:
            g.drawImage(this.casePic[0], c.x, c.y);
            break;
          case 1:
            g.drawImage(this.casePic[1], c.x, c.y);
            g.drawImage(this.itemPic[this.boxID[i][j]], c.x + 10, c.y);
            if (this.numPRINT === 0 && this.PRINTed[i][j] === 0) {
              this.PRINTed[i][j] = 1;
            } else if (this.numPRINT === 1 && this.PRINTed[i][j] === 0) {
              this.PRINTed[i][j] = 2;
            }

            if (this.PRINTed[i][j] === 1) {
              this.drawText(g, this.names[this.boxID[i][j]], { x: 25, y: 25 },
                            KLevel.RIGHT, KLevel.TOP, [255, 255, 255], false);
              this.numPRINT = 1;
            } else if (this.PRINTed[i][j] === 2) {
              this.drawText(g, "und", { x: 200, y: 25 },
                            KLevel.CENTER, KLevel.TOP, [255, 255, 255], false);
              this.drawText(g, this.names[this.boxID[i][j]], { x: 375, y: 25 },
                            KLevel.LEFT, KLevel.TOP, [255, 255, 255], false);
              this.numPRINT = 0;
            }
            break;
          case 2:
            this.numMatched++;
            break;
          case 3:
            this.drawText(g, "ein Diamant", { x: 200, y: 75 },
                          KLevel.CENTER, KLevel.TOP, [0, 255, 255], false);
            g.drawImage(this.casePic[1], c.x, c.y);
            g.drawImage(this.itemPic[0], c.x + 10, c.y);
            this.boxST[i][j] = 2;
            this.DiamantSound.play();
            this.enableNextLevel();
            this.changeScore(100);
            this.timOn = false;
            break;
        }
      }
    }

    if (this.numOpen >= 2) {
      this.numOpen = 0;
      this.timOn = false;
      this.checkMatch(g, this.curOpen[0].x, this.curOpen[0].y,
                      this.curOpen[1].x, this.curOpen[1].y);
    }

    if (this.timOn && this.numMatched < 15) {
      this.timOn = false;
      this.stop();
    } else if (this.numMatched >= 15) {
      this.displayText(g, "Next Level: +100pts");
      this.changeScore(100);
      this.stop();
      this.numOpen = 0;
    } else {
      this.timOn = true;
    }
  }

  mouseDown(x, y) {
    for (let i = 0; i < 5; i++) {
      for (let j = 0; j < 3; j++) {
        const c = this.content[i][j];
        if (x >= c.x && x < c.x + c.w && y >= c.y && y < c.y + c.h &&
            this.boxST[i][j] === 0) {
          this.boxST[i][j] = 1;
          if (this.boxID[i][j] === 0) {
            this.boxST[i][j] = 3;
          } else {
            this.numOpen++;
            this.curOpen[this.numOpen - 1] = { x: i, y: j };
          }
          this.timOn = false;
          this.start();
        }
      }
    }
  }

  checkMatch(g, x1, y1, x2, y2) {
    if (this.boxID[x1][y1] === this.boxID[x2][y2]) {
      this.drawText(g, "Ja", { x: 200, y: 75 },
                    KLevel.CENTER, KLevel.TOP, [255, 255, 0], false);
      this.Ja.play();
      this.boxST[x1][y1] = 2;
      this.boxST[x2][y2] = 2;
      this.timOn = false;
      this.changeScore(50);
    } else {
      this.boxST[x1][y1] = 0;
      this.boxST[x2][y2] = 0;
      this.drawText(g, "Nein", { x: 200, y: 75 },
                    KLevel.CENTER, KLevel.TOP, [255, 255, 0], false);
      this.Nein.play();
      this.timOn = false;
      this.changeScore(-10);
    }

    this.PRINTed[x1][y1] = 0;
    this.PRINTed[x2][y2] = 0;
    this.numPRINT = 0;
  }
}
